### PR TITLE
Fireperf EAP: custom `cold_start` and `prewarmed_start` trace for prewarming identification protytpe

### PR DESCRIPTION
**Pull request by leotianlizhan**
b/211905766

### CODE REVIEW

File: FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m [hunk 1]
```diff
@@ -143,26 +147,38 @@ - (FIRTrace *)activeTrace {
...
+  if (![self prewarmAvailable]) {
     return NO;
   }
 
-  NSDictionary<NSString *, NSString *> *environment = [NSProcessInfo processInfo].environment;
-  if (environment[@"ActivePrewarm"] != nil &&
-      [environment[@"ActivePrewarm"] isEqualToString:@"1"]) {
+  if ([self isActivePrewarm]) {
     return YES;
   }
```
Comment (jeremyjiang-dev) [inline FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:157]: Nit: `return [self isActivePrewarm] || [self isDoubleDispatchPrewarm]`

File: FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m [hunk 2]
```diff
@@ -221,6 +248,29 @@ - (void)appDidBecomeActiveNotification:(NSNotification *)notification {
...
         [self.appStartTrace cancel];
       }
+
+      if ([self prewarmAvailable] &&
+          (currentTimeSinceEpoch - startTimeSinceEpoch < gAppStartMaxValidDuration)) {
+        BOOL activePrewarm = [self isActivePrewarm];
+        BOOL doubleDispatchPrewarm = [self isDoubleDispatchPrewarm];
+        if (!activePrewarm && !doubleDispatchPrewarm) {
+          [self.coldStartTrace stop];
+          [self.prewarmedStartTrace cancel];
```
Comment (jeremyjiang-dev) [inline FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:258]: Why do you need `return`? 
Nit: Can you connect `if (!activePrewarm && !doubleDispatchPrewarm)` and `if (activePrewarm && doubleDispatchPrewarm)` with `else if` for better readability?

**Review by jeremyjiang-dev — COMMENTED** [review]
The changes LGTM! Can you add some tests?

**Review by jeremyjiang-dev — APPROVED** [review]

**Comment by leotianlizhan** [comment]
> We also need to update [the SDK version](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebasePerformance.podspec#L3) to something like `ios-15-eap` in this branch.

does the SDK version have to follow a number pattern like `8.12.0` or does it work with any string like `ios-15-eap`?

**Comment by jeremyjiang-dev** [comment]
> > We also need to update [the SDK version](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebasePerformance.podspec#L3) to something like `ios-15-eap` in this branch.
> 
> does the SDK version have to follow a number pattern like `8.12.0` or does it work with any string like `ios-15-eap`?

Let's add Paul to review the PR so that we are following the best practice.

File: FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m [hunk 3]
```diff
@@ -143,24 +146,49 @@ - (FIRTrace *)activeTrace {
...
+    [trace setValue:@"active_prewarm" forAttribute:@"prewarm_detection"];
+    [trace setValue:@"yes" forAttribute:@"is_prewarmed"];
+  } else {
+    [trace setValue:@"double_dispatch" forAttribute:@"prewarm_detection"];
+    [trace setValue:@"yes" forAttribute:@"is_prewarmed"];
   }
+}
+
+- (BOOL)prewarmAvailable {
+  NSArray *versionComponents =
```
Comment (paulb777) [inline FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:178]: May also check that the platform is iOS. I'm not aware of the prewarm issue is also on tvOS. Also consider using https://github.com/google/GoogleUtilities/blob/main/GoogleUtilities/Environment/third_party/GULAppEnvironmentUtil.m#L245 to get the OS version

File: FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m [hunk 4]
```diff
@@ -143,24 +146,49 @@ - (FIRTrace *)activeTrace {
...
+      [[UIDevice currentDevice].systemVersion componentsSeparatedByString:@"."];
+  NSInteger majorVersion = [versionComponents[0] integerValue];
+  return majorVersion >= 15;
+}
+
+- (BOOL)isActivePrewarm {
+  NSDictionary<NSString *, NSString *> *environment = [NSProcessInfo processInfo].environment;
+  return (environment[@"ActivePrewarm"] != nil &&
```
Comment (paulb777) [inline FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:186]: May want to use a plist value to check for enabling the EAP code to make it easy to include in a release and avoid separate release management for the EAP.  For example, see https://github.com/firebase/firebase-ios-sdk/issues/8949#issuecomment-975970983

Comment (leotianlizhan) [reply FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:186]: Do you mean we make a normal release, then ask EAP customers to set a plist value to enable this prototype?

Comment (paulb777) [reply FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:186]: Yes. It's easier from a documentation/release management/installation/usability perspective.

Comment (paulb777) [reply FirebasePerformance/Sources/AppActivity/FPRAppActivityTracker.m:193]: Discussed offline and agreed this is appropriate for a limited EAP

**Comment by leotianlizhan** [comment]
Since there has been no signups for EAP, this branch will be closed.

**Pull request closed without merging**